packages/ts/models/src/builders.ts

```diff
@@ -0,0 +1,254 @@
+ * @typeParam F - The flags for the model constructor that allow to determine
+ * specific characteristics of the model.
+ *
+ * @param base - The base model to extend.
+ * @param defaultValueProvider - The function that provides the default value
+ * for the model.
+ *
+ * @internal
+ */
+export class CoreModelBuilder<
+  V,
+  EX extends AnyObject = EmptyObject,
+  F extends Flags = { named: false; selfRefKeys: never },
+> {
+  protected readonly [$model]: Model<V, EX>;
+
+  constructor(base: Model, defaultValueProvider?: (model: Model<V, EX>) => V) {
+    this[$model] = create(base);
+
+    if (defaultValueProvider) {
+      this.defaultValueProvider(defaultValueProvider);
+    }
+  }
+
+  /**
+   * Appends metadata to the model.
+   *
+   * @param value - The metadata to append.
+   * @returns The current builder instance.
+   */
+  meta(value: ModelMetadata): this {
+    this.define($meta, { value });
+    return this;
+  }
+
+  /**
+   * Defines a new property on the model. The property serves the purposes of storing the extra data for specific types
+   * of models.
+   *
+   * @remarks
+   * The key of the property should be a symbol to avoid conflicts with properties defined via
+   * {@link ObjectModelBuilder}.
+   *
+   * @param key - The key of the property.
+   * @param value - The descriptor of the property.
+   * @returns The current builder instance.
+   */
+  define<DK extends symbol, DV>(
+    key: DK,
+    value: TypedPropertyDescriptor<DV>,
+  ): CoreModelBuilder<V, EX & Readonly<Record<DK, DV>>, F> {
+    defineProperty(this[$model], key, value);
+    return this as any;
+  }
+
+  /**
+   * Sets the default value provider for the model. This is an alternative way
+   * to provide the default value for the model if for some reason using the
+   * constructor parameter is undesired.
+   *
+   * @param defaultValueProvider - The function that provides the default value for the
+   * model.
+   * @returns The current builder instance.
+   */
+  defaultValueProvider(defaultValueProvider: DefaultValueProvider<V, EX>): this {
+    this.define($defaultValue, {
+      get(this: Model<V, EX>) {
+        return defaultValueProvider(this);
+      },
+    });
+    return this;
+  }
+
+  /**
+   * Sets the name of the model. The name is used for debugging purposes and is
+   * displayed in the string representation. Setting the name is required;
+   * otherwise, the {@link CoreModelBuilder.build} method won't be available.
+   *
+   * @param name - The name of the model.
+   * @returns The current builder instance.
+   */
+  name(name: string): CoreModelBuilder<V, EX, { named: true; selfRefKeys: F['selfRefKeys'] }> {
+    return this.define($name, { value: name }) as any;
+  }
+
+  /**
+   * Builds the model.
+   *
+   * @returns The model.
+   */
+  build(this: F['named'] extends true ? this : never): Model<V, EX> {
+    return this[$model];
+  }
+}
+
+const propertyRegistry = new WeakMap<Model, Record<string, Model>>();
```

Comment:
You mean the explanation why `WeakMap` is used over `Map`, or a description what this registry does?

Reply:
I mean why a `WeakMap` over a `Map`. I know the difference in how the references are stored, yet the reason of the choice is not clear to me.

Reply:
Well, it's kind of a rule of thumb: if you have an object as a key and you don't plan to iterate over the map, use `WeakMap`, just in case any of the object is garbage collected. I'm not sure how to explain it in the comment 🤔

Reply:
But what does the `propertyRegistry` hold exactly? By looking at the code, it seems like it holds some objects that are empty at the beginning and properties are added to them each time. If that's the case and you leave them available for garbage collection, there is a possibility that they are recreated and start again from an empty state. Are they referenced elsewhere? This is the kind of reasoning that I would like to see commented here.

Reply:
Well, this registry is basically a some kind of a private property list for a specific model (the key object). So, these properties exist only if the model they are bound to exists. That's why `WeakMap`. So, no, they won't be recreated again if the model is garbage collected.

Reply:
I can put this explanation as a comment for the registry btw.